File: tests/smooth.test.ts
import { describe, expect, it } from 'vitest';
import type { TrackPoint } from '../src/state.js';
import {
  douglasPeucker,
  jumpSplit,
  movingAverage,
  smoothTrack,
} from '../src/smooth.js';

const pt = (
  lat: number,
  lng: number,
  t: number,
  jump?: boolean,
): TrackPoint => ({ t, pos: { lat, lng }, acc: 5, seg: 's0', jump });

describe('jumpSplit', () => {
  it('无 jump 为单段', () => {
    const segs = jumpSplit([pt(0, 0, 0), pt(1, 1, 1), pt(2, 2, 2)]);
    expect(segs).toHaveLength(1);
    expect(segs[0]).toHaveLength(3);
  });

  it('中间 jump 切两段，jump 点归后段', () => {
    const segs = jumpSplit([
      pt(0, 0, 0),
      pt(1, 1, 1),
      pt(2, 2, 2, true),
      pt(3, 3, 3),
    ]);
    expect(segs).toHaveLength(2);
    expect(segs[0].map((p) => p.pos.lat)).toEqual([0, 1]);
    expect(segs[1].map((p) => p.pos.lat)).toEqual([2, 3]);
  });

  it('开头 jump 不产生空段', () => {
    const segs = jumpSplit([pt(0, 0, 0, true), pt(1, 1, 1)]);
    expect(segs).toHaveLength(1);
    expect(segs[0]).toHaveLength(2);
  });
});

describe('movingAverage', () => {
  it('长度不变、保留 t/acc/jump', () => {
    const pts = [pt(0, 0, 0), pt(1, 1, 1, true), pt(2, 2, 2)];
    const out = movingAverage(pts);
    expect(out).toHaveLength(3);
    expect(out[1].t).toBe(1);
    expect(out[1].jump).toBe(true);
  });

  it('共线等距点不变（窗口均值等于自身）', () => {
    const pts = [pt(0, 0, 0), pt(0.001, 0, 1), pt(0.002, 0, 2), pt(0.003, 0, 3), pt(0.004, 0, 4)];
    const out = movingAverage(pts);
    expect(out[2].pos.lat).toBeCloseTo(0.002, 9);
  });

  it('端点窗口收窄取均值', () => {
    const pts = [pt(0, 0, 0), pt(2, 0, 1), pt(4, 0, 2)];
    const out = movingAverage(pts);
    // w=5, half=2：首点窗口 [0,2] → lat 均值 2
    expect(out[0].pos.lat).toBeCloseTo(2, 9);
    expect(out[2].pos.lat).toBeCloseTo(2, 9);
  });
});

describe('smoothTrack', () => {
  it('长度不变', () => {
    const pts = [pt(0, 0, 0), pt(1, 1, 1, true), pt(2, 2, 2), pt(3, 3, 3)];
    expect(smoothTrack(pts)).toHaveLength(4);
  });

  it('窗口不跨 jump 段', () => {
    const farA = pt(0, 0, 0);
    const farB = pt(1, 1, 1, true); // 相距 ~157km 的跳变
    const farC = pt(2, 2, 2);
    const out = smoothTrack([farA, farB, farC]);
    // farB 在独立段，其坐标不受 farA/farC 拉拽
    expect(out[1].pos).toEqual({ lat: 1, lng: 1 });
    expect(out[0].pos).toEqual({ lat: 0, lng: 0 });
  });
});

describe('douglasPeucker', () => {
  it('直线仅保留两端点', () => {
    const line = [
      { lat: 0, lng: 0 },
      { lat: 1, lng: 1 },
      { lat: 2, lng: 2 },
      { lat: 3, lng: 3 },
    ];
    const out = douglasPeucker(line, 2);
    expect(out).toEqual([
      { lat: 0, lng: 0 },
      { lat: 3, lng: 3 },
    ]);
  });

  it('eps=0 全保留', () => {
    const line = [
      { lat: 0, lng: 0 },
      { lat: 1, lng: 1.5 },
      { lat: 2, lng: 2 },
    ];
    expect(douglasPeucker(line, 0)).toHaveLength(3);
  });

  it('直角拐点保留（拐点距离大于 eps）', () => {
    const corner = [
      { lat: 0, lng: 0 },
      { lat: 1, lng: 0 }, // 拐点
      { lat: 1, lng: 1 },
    ];
    expect(douglasPeucker(corner, 1)).toEqual(corner);
  });

  it('空与两点直通', () => {
    expect(douglasPeucker([], 2)).toEqual([]);
    const two = [
      { lat: 0, lng: 0 },
      { lat: 1, lng: 1 },
    ];
    expect(douglasPeucker(two, 2)).toEqual(two);
  });
});

File: android/core/src/test/kotlin/io/github/chenchen913/baibai/core/SmoothTest.kt
package io.github.chenchen913.baibai.core

import io.github.chenchen913.baibai.core.model.LatLng
import io.github.chenchen913.baibai.core.model.TrackPoint
import io.github.chenchen913.baibai.core.smooth.Smooth
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/** 对应网页版 tests/smooth.test.ts（12 项，逐位对齐） */
class SmoothTest {

    private fun pt(lat: Double, lng: Double, t: Long, jump: Boolean? = null): TrackPoint =
        TrackPoint(t = t, pos = LatLng(lat, lng), acc = 5.0, seg = "s0", jump = jump)

    // ---------- jumpSplit ----------

    @Test
    fun `jumpSplit 无 jump 为单段`() {
        val segs = Smooth.jumpSplit(listOf(pt(0.0, 0.0, 0), pt(1.0, 1.0, 1), pt(2.0, 2.0, 2)))
        assertEquals(1, segs.size)
        assertEquals(3, segs[0].size)
    }

    @Test
    fun `jumpSplit 中间 jump 切两段，jump 点归后段`() {
        val segs = Smooth.jumpSplit(
            listOf(pt(0.0, 0.0, 0), pt(1.0, 1.0, 1), pt(2.0, 2.0, 2, jump = true), pt(3.0, 3.0, 3)),
        )
        assertEquals(2, segs.size)
        assertEquals(listOf(0.0, 1.0), segs[0].map { it.pos.lat })
        assertEquals(listOf(2.0, 3.0), segs[1].map { it.pos.lat })
    }

    @Test
    fun `jumpSplit 开头 jump 不产生空段`() {
        val segs = Smooth.jumpSplit(listOf(pt(0.0, 0.0, 0, jump = true), pt(1.0, 1.0, 1)))
        assertEquals(1, segs.size)
        assertEquals(2, segs[0].size)
    }

    // ---------- movingAverage ----------

    @Test
    fun `movingAverage 长度不变、保留元字段`() {
        val pts = listOf(pt(0.0, 0.0, 0), pt(1.0, 1.0, 1, jump = true), pt(2.0, 2.0, 2))
        val out = Smooth.movingAverage(pts)
        assertEquals(3, out.size)
        assertEquals(1L, out[1].t)
        assertEquals(true, out[1].jump)
    }

    @Test
    fun `movingAverage 共线等距点不变`() {
        val pts = listOf(
            pt(0.0, 0.0, 0),
            pt(0.001, 0.0, 1),
            pt(0.002, 0.0, 2),
            pt(0.003, 0.0, 3),
            pt(0.004, 0.0, 4),
        )
        val out = Smooth.movingAverage(pts)
        assertEquals(0.002, out[2].pos.lat, 1e-9)
    }

    @Test
    fun `movingAverage 端点保持原坐标，中间点取窗口均值`() {
        val pts = listOf(pt(0.0, 0.0, 0), pt(1.0, 0.0, 1), pt(2.0, 0.0, 2), pt(3.0, 0.0, 3), pt(4.0, 0.0, 4))
        val out = Smooth.movingAverage(pts)
        assertEquals(0.0, out[0].pos.lat)
        assertEquals(4.0, out[4].pos.lat)
        assertEquals(2.0, out[2].pos.lat, 1e-9)
    }

    // ---------- smoothTrack ----------

    @Test
    fun `smoothTrack 长度不变`() {
        val pts = listOf(pt(0.0, 0.0, 0), pt(1.0, 1.0, 1, jump = true), pt(2.0, 2.0, 2), pt(3.0, 3.0, 3))
        assertEquals(4, Smooth.smoothTrack(pts).size)
    }

    @Test
    fun `smoothTrack 窗口不跨 jump 段`() {
        val farA = pt(0.0, 0.0, 0)
        val farB = pt(1.0, 1.0, 1, jump = true)
        val farC = pt(2.0, 2.0, 2)
        val out = Smooth.smoothTrack(listOf(farA, farB, farC))
        assertEquals(LatLng(1.0, 1.0), out[1].pos)
        assertEquals(LatLng(0.0, 0.0), out[0].pos)
    }

    // ---------- douglasPeucker ----------

    @Test
    fun `douglasPeucker 直线仅保留两端点`() {
        val line = listOf(LatLng(0.0, 0.0), LatLng(1.0, 1.0), LatLng(2.0, 2.0), LatLng(3.0, 3.0))
        assertEquals(listOf(LatLng(0.0, 0.0), LatLng(3.0, 3.0)), Smooth.douglasPeucker(line, 2.0))
    }

    @Test
    fun `douglasPeucker eps 为 0 全保留`() {
        val line = listOf(LatLng(0.0, 0.0), LatLng(1.0, 1.5), LatLng(2.0, 2.0))
        assertEquals(3, Smooth.douglasPeucker(line, 0.0).size)
    }

    @Test
    fun `douglasPeucker 直角拐点保留`() {
        val corner = listOf(LatLng(0.0, 0.0), LatLng(1.0, 0.0), LatLng(1.0, 1.0))
        assertEquals(corner, Smooth.douglasPeucker(corner, 1.0))
    }

    @Test
    fun `douglasPeucker 空与两点直通`() {
        assertEquals(emptyList<LatLng>(), Smooth.douglasPeucker(emptyList(), 2.0))
        val two = listOf(LatLng(0.0, 0.0), LatLng(1.0, 1.0))
        assertEquals(two, Smooth.douglasPeucker(two, 2.0))
        assertTrue(true)
    }
}
